### TF-IDF

tf-idf（英语：term frequency–inverse document frequency）是一种用于信息检索与文本挖掘的常用加权技术。tf-idf是一种统计方法，用以评估一字词对于一个文件集或一个语料库中的其中一份文件的重要程度。字词的重要性随着它在文件中出现的次数成正比增加，但同时会随着它在语料库中出现的频率成反比下降。

In [34]:
import pandas as pd
import jieba
import time
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics
from sklearn.externals import joblib
from sklearn.utils import shuffle

In [21]:
def chinese_word_cut(s):
    # 中文分词（jieba）
    return ' '.join(jieba.cut(s))


def train_and_predict(train_x, train_y, test_x, test_y):
    # 使用多项分布朴素贝叶斯分类器进行训练
    clf = MultinomialNB().fit(train_x, train_y)
    # 在本地保存 model 参数
    joblib.dump(clf, 'model.pkl')
    # 预测 test_x 中的结果
    predicted = clf.predict(test_x)
    # 打印报表
    print(metrics.classification_report(test_y, predicted))
    print('accuracy_score: {}'.format(metrics.accuracy_score(test_y, predicted)))


def read_vocab(path):
    # 读取 path 下文本的内容，返回 list https://github.com/goto456/stopwords
    f = open(path, 'r', encoding='utf8')
    res = f.readlines()
    f.close()
    return res

### 加载训练集以及测试集

In [35]:
train_data = pd.read_csv('cnews.train.txt', sep='\t', names=['label', 'content'])
test_data = pd.read_csv('cnews.test.txt', sep='\t', names=['label', 'content'])
# shuffle 一下
shuffle(train_data)
shuffle(test_data)

label                                            content
9471    财经  中银基金公司获批QDII资格本报讯 记者从中银基金管理公司获悉，该公司已正式获得合格境内机构...
2260    家居  品种不同的地板 产生缝隙的原因也不同范小姐家的竹地板出现了不明原因的裂缝。读者范小姐致电京华...
530     体育  调查-如何评价詹姆斯34分热火末节惊魂险胜老鹰？新浪体育讯从领先20分，到被对手追平，再到拿...
962     体育  76人vs热火V前瞻：费城破釜沉舟 斯帅临阵换将？新浪体育讯尽管输掉了第四场比赛，但对于总比...
6942    时政  部委部署铁路民警向公务员过渡工作中新网7月28日电 国家公务员局网站日前发布消息称，国家公务...
9095    财经  我爱你基金定投亮相在情人节档期，继“我爱宝贝”基金定投计划推出之后，易方达基金近期针对情侣、...
3412    房产  一线城市楼价飙得高 更易受秋凉 记者走访北京、上海、广州、深圳等地发现，部分一线城市的楼市价...
3869    房产  1月全国房价同比上涨9.5%本报讯(记者孟为)国家发改委、国家统计局昨天公布调查显示，201...
408     体育  山猫vs热火首发：韦德复出回归首发 山猫群龙无首新浪体育讯北京时间4月9日，迈阿密热火坐镇主...
3267    房产  金九成色不足银十变铜十 楼市走向有些看不懂今年楼市的走向，越来越让人难以啄摸，跟运行了十几年...
2017    家居  促销带动消费陶瓷卫浴品牌要再升级促销可以推动产品的销售范围和刺激消费者的购买意识。 促销战拉...
8808    科技  纤薄时尚 尼康S220售1299元还有好礼相送S220是尼康09年S系列新品相机中最薄一款，...
8073    科技  热门高端智能卡片 索尼TX1套装带票促销编辑观点：索尼T系列卡片机一直都是相机市场的明星产品...
237     体育  江苏客场不敌吉林被双杀 胡雪峰砍23+7+8却难救主新浪体育讯北京时间2月11日，2010-...
682     体育  大唐受伤南钢告急换外援 玻璃人霍斯金将取代莱特一场在奥体的救赎之战，并不能挽救南钢外援莱特被...
4950    教育  经验分享：拼搏两个月英语六级拿高分为了对付今年6月份的六级，我苦苦拼搏了整整两个月，出奇的得...
432     体育  科比：四连败只因疲惫 禅师暴怒：你们根本没想赢!“球迷们在高呼他的名字，我也开始加入到其中，...
8296    科技  机身防抖翻转屏幕 奥巴E620双头套5550    作者：陈亮    在设计E-620的图像...
8925    科技  轻薄防抖时尚卡片DC 松下FS6仅售1000元松下 FS6采用了810万有效像素1/2.5英...
6445    时政  胡锦涛抵达斯洛伐克开始国事访问新华网布拉迪斯拉发6月18日电 中国国家主席胡锦涛18日抵达布...
9042    财经  嘉实量化阿尔法基金获批嘉实量化阿尔法股票型证券投资基金已于日前获得中国证监会核准，将于近期发...
4226    教育  中国在美EB2类移民截止日排期或小幅前移据美国《侨报》讯 数周前，律师参加了在圣地亚哥举行的...
1840    娱乐  组图：宁浩客串《刀见笑》 斗鸡客造型喜感十足新浪娱乐讯 由乌尔善导演，张雨绮、安藤政信和游本...
3487    房产  楼盘成本清单引发强烈反响 谁是高房价受损者本报10月27日二版刊登的《一份楼盘“成本清单”揭...
5666    时尚  组图：想性感 露得学问露得漂亮导语：凹凸有致一贯是展现女性美最好的武器。无限低开的深V形领，...
4718    教育  准留学生教育展后必做3件事以便明确专业方向北京考试报讯  500多所国外院校参加的大型国际教...
8666    科技  60倍光变硬盘DV！索尼SR47E单机2280【山东IT在线报道】索尼作为最新推出的数码摄像...
7661    游戏  《明珠幻想》7日火爆内测 狂送内测号2009年末，最绚丽的手机网游巨作，北京掌上明珠信息技术...
8394    科技  实用更廉价 索尼α350单反仅售3999元【云南IT我帮您 www.ynit580.com】...
3393    房产  通胀预期下房产投资有两大风险尹伯成(复旦大学房地产研究中心主任)：时下，为规避通胀风险而进行...
...    ...                                                ...
4232    教育  世界各国公务员千奇百怪的办公室(组图)新民网讯：“公务员(微博)”一词，由英语“civils...
4760    教育  ETS最新消息：托福阅读11月起有“三变”北京考试报讯(记者 马海丽) 美国教育考试服务中心...
9298    财经  谨慎关注封基交易性机会□天相投顾 闻群 王聃聃2009年以来在以投资为主导的经济刺激政策作用...
8954    科技  超薄时尚卡片机 卡西欧Z280套装售价1320元卡西欧Z280做工精致出色，整体设计简约并散...
5344    时尚  组图：吊袜带 诠释性感的最佳单品导语：丝袜，诠释性感的最佳单品。对于丝袜你最中意哪款？短口丝...
4764    教育  忧经济不敢多收人 加拿大明年移民额维持25万中新网(微博)10月30日电 据加拿大《明报》报...
5094    时尚  Marc by Marc Jacobs女孩夏天穿JellyJelly今季的胶鞋一街都是，平、...
6923    时政  香港元朗发生严重车祸致3死14伤(图)中新网7月26日电 据香港《文汇报》报道，香港元朗公路...
3187    房产  深圳楼市金九落空 门前冷落车马稀开发商期待中的“金九”并没有如期而至。九月份深圳楼市新房成交...
5003    时尚  魅惑民族风情透出帅气今夏，女孩们尽可挥洒摇滚、波西米亚、魅惑和民族风情，如牛仔热裤或花纹紧腿...
5263    时尚  组图：明星沙滩比基尼 是秀场还是失地导语：这年头，宁可相信男人，也不要信大明星的宣传照，靠着...
2826    家居  掌握夏季家居清洁关键点 健康一夏轻松度过关键点1 ——空调炎炎夏日，空调是人们居家生活不可缺...
4088    教育  加拿大留学顺利获签三大要点解读经济能力是通过签证的首要条件。由于个别申请人在申请时出现弄虚作...
8809    科技  轻薄时尚卡片DC 佳能IXUS100 IS仅1650说佳能IXUS100拥有“倾城之貌”并不...
7669    游戏  《幻想i时代》精宠攻略全揭秘为了满足广大《幻想i时代》友友们对精宠的需求，今天写一点自己的经...
2782    家居  惠达王彦庆赢取更多资源 优化媒体平台(图)     新浪家居谢娟讯  10月16日，易居中国...
8635    科技  4倍光学变焦镜头 索尼W220仅售1288元作者：杨子毅索尼W220使用了一个2.7英寸、2...
7070    游戏  云光仙境赤月战争 游戏各族简要攻略*注意：本游戏很多地方考验玩家的策略以及思考能力，游戏过程...
5549    时尚  黑色的长裙显得稳重妩媚不是碎花，也不是白色，黑色的长裙显得稳重妩媚，配个小外套很实用双排扣短...
6108    时政  民政部：去年民政系统彩票公益金支出119亿新华网北京5月24日电(记者卫敏丽)民政部日前发布...
1417    娱乐  《倾城之泪》亮相合肥 开“泪闸”净化观影心灵新浪娱乐讯 据悉，2011年度爱情催泪巨制《倾城...
5693    时尚  组图：超Q小内裤 穿高跟鞋内衬小物导语：猜猜看，这是什么？其实这是为舞蹈演员和运动员双脚设计...
7390    游戏  《幻想i时代》公会岛屿 激烈争夺10月30日，北京掌上明珠信息技术有限公司大型魔幻题材手机网...
4315    教育  欧盟驻华大使：欧盟移民政策暂不会改变(图)新快报讯  为了悼念爆炸和枪击事件中的死难者，挪威...
4344    教育  大马非法外劳漂白计划实施一周逾24万人登记中新网8月6日电 据新加坡《联合早报》报道，马来西...
5887    时尚  组图：寒冬不无聊 教你玩转配饰新花样导语：沉闷的冬季着装的亮点在于小小的配饰，一件适合的漂亮...
306     体育  杜兰特32分雷霆胜掘金 大卫李29+20勇士灭开拓者新浪体育讯北京时间4月6日，雷霆客场击败...
3419    房产  中国地产金融年会互动演讲：国进是否等于民退10月15日：    大变局大应对第五届中国地产金...
489     体育  绿军更衣室竟推迟半小时才打开 皮尔斯轰记者离场新浪体育讯北京时间4月11日(迈阿密时间4月1...
1347    娱乐  王小帅《我11》釜山首映 称审查已非国产片难题新浪娱乐讯 10月8日上午，王小帅导演的新作《...

[10000 rows x 2 columns]

### 对于数据集中每一个句子进行中文分词

In [43]:
train_data['content'] = train_data['content'].apply(chinese_word_cut)
test_data['content'] = test_data['content'].apply(chinese_word_cut)

### 根据词袋向量统计TF-IDF

定义 TfidfVectorizer，将 train_data 以及 test_data 转换为词向量并计算 TF-IDF

In [44]:
# 可以直接在 TfidfVectorizer 中传入停用词，设置词汇维度最大为 max_features
tfidf = TfidfVectorizer(max_features=100000, stop_words=read_vocab('stop_word.txt'))
# 拼接 train_data 以及 test_data 的作用是为了获得同一空间中的词汇向量（但实际中还是提前使用训练集做个映射表比较好，因为测试集中可能某些词汇在训练集中没有出现过）
# 为了偷懒一起做方便很多
x = tfidf.fit_transform(train_data['content'].append(test_data['content']))
train_x = x[:len(train_data)]
test_x = x[len(train_data):]
train_y = train_data['label']
test_y = test_data['label']

C:\Anaconda3\envs\mLearning\lib\site-packages\sklearn\feature_extraction\text.py:300: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['一些', '一何', '一切', '一则', '一方面', '一旦', '一来', '一样', '一般', '一转眼', '万一', '上下', '不仅', '不但', '不光', '不单', '不只', '不外乎', '不如', '不妨', '不尽', '不尽然', '不得', '不怕', '不惟', '不成', '不拘', '不料', '不是', '不比', '不然', '不特', '不独', '不管', '不至于', '不若', '不论', '不过', '不问', '与其', '与其说', '与否', '与此同时', '且不说', '且说', '两者', '个别', '为了', '为什么', '为何', '为止', '为此', '为着', '乃至', '乃至于', '之一', '之所以', '之类', '乌乎', '也好', '也罢', '二来', '于是', '于是乎', '云云', '云尔', '人们', '人家', '什么', '什么样', '介于', '仍旧', '从此', '从而', '他人', '他们', '以上', '以为', '以便', '以免', '以及', '以故', '以期', '以来', '以至', '以至于', '以致', '任何', '任凭', '似的', '但凡', '但是', '何以', '何况', '何处', '何时', '余外', '作为', '你们', '使得', '例如', '依据', '依照', '便于', '俺们', '倘使', '倘或', '倘然', '倘若', '假使', '假如', '假若', '傥然', '先不先', '光是', '全体', '全部', '关于', '其一', '其中', '其二', '其他', '其余', '其它', '其次', '具体地说', '具体说来', '兼之', '再其次', '

### 测试文章中词语的 TF-IDF 权重与文章类别之间的关系

抽取 5 条数据分别计算其中词语的 TF-IDF 权重并降序输出前 10 个关键词，并与其真实类别对比

tfidf.get_feature_names 可以获取当前模型中所有的词语，最大数量为 max_features

In [54]:
word = tfidf.get_feature_names()
for i in np.random.randint(train_x.shape[0], size=5):
    print('label: ', train_y[i])
    arg_sort = np.argsort(-train_x[i].toarray()[0])[:10]
    for j in arg_sort:
        print(word[j], '\t', train_x[i].toarray()[0][j])
    print('--------------------------------')

label:  房产
流动人口 	 0.4082248955538784
计划生育 	 0.34829150970183487
改革 	 0.34125189927858635
深化 	 0.19670126794087855
资本金 	 0.19556879143230255
推进 	 0.1890876174017774
项目 	 0.15657100082162545
条例 	 0.1437456888734804
会议 	 0.13021221269175748
投资 	 0.12625010914409945
--------------------------------
label:  娱乐
激浪 	 0.4316732046209529
青春 	 0.3830683472008551
黄轩 	 0.3179488217376342
黄晓明 	 0.2286114607959972
龙舟 	 0.18052996401984894
陈乔恩 	 0.17527095996307465
梁柏坚 	 0.13644213118630136
车震 	 0.13438219859039577
法宝 	 0.1238574067351264
青春片 	 0.12252506778901526
--------------------------------
label:  教育
雅思考试 	 0.5728694476032157
考生 	 0.2701400489157494
史哲明 	 0.24672257560968106
雅思 	 0.22926765074731068
文化教育 	 0.17088583079652214
大使馆 	 0.15683943946480075
院校 	 0.1563748289754174
超过 	 0.14911140474311899
认可 	 0.14811921410744747
英国 	 0.12173639512952922
--------------------------------
label:  房产
农村 	 0.4475497399879565
吕祖善 	 0.35359914278201837
农民 	 0.3179448351651087
改造 	 0.26207820762374934
住房 	

### 训练并预测

In [48]:
train_and_predict(train_x, train_y, test_x, test_y)

              precision    recall  f1-score   support

          体育       1.00      1.00      1.00      1000
          娱乐       0.93      0.99      0.96      1000
          家居       0.97      0.38      0.55      1000
          房产       0.62      0.92      0.74      1000
          教育       0.91      0.94      0.93      1000
          时尚       0.97      0.97      0.97      1000
          时政       0.95      0.91      0.93      1000
          游戏       0.98      0.97      0.97      1000
          科技       0.95      0.99      0.97      1000
          财经       0.95      0.99      0.97      1000

    accuracy                           0.91     10000
   macro avg       0.92      0.91      0.90     10000
weighted avg       0.92      0.91      0.90     10000

accuracy_score: 0.906


## 总结

本次实验使用 jieba 对原始数据中的文本进行分词，然后利用 TfidfVectorizer 对语料集构建了所有词语的映射，再通过计算 TF-IDF，从人工的角度观察发现一篇文章中排名前 10 的关键词与其类别关联度很大。

最后使用朴素贝叶斯的方法使用 train_data 进行训练，随后在 test_data 中进行测试，得到的最终准确率为 90.6%，其结果也是相当可观的。

从结果上看，对“体育”的分类准确率高达 100%，其次是“游戏”，随意抽取几个体育和游戏类文章的关键字来看，其词汇相对于其他类别来说更加明显。

如“体育”类别：布鲁诺、凯莉、红地毯、马斯、全明星赛，基本上很多都是著名的体育明星。

如“游戏”类别：地图、星际、rpg、玩家、dota